### PR TITLE
lint: Deny #[no_mangle] const items

### LINKED ISSUE #21856: pub const items have no linker symbol
```rust
#[no_mangle]
pub const THING: libc::c_int = 42;
```

Will not actually be exported in the resulting object. There are two obvious fixes, the first being to expand on my pub-not-exported lint to warn you that this will not work, the second being to fix the compiler to export the symbol with the appropriate value (Without altering the existing inlining behaviour, although it may give a more reasonable option for cases like trying to take the address of a `const`item).

Happy to work on either, although I would need a mentor for the second part.

### PR DESCRIPTION
**Pull request by richo**
This renames the PrivateNoMangleFns lint to allow both to happen in a
single pass, since they do roughly the same work.

Closes #21856

Open questions:

[ ]: Do the tests actually pass (I'm running make check and running out the door now)
[ ]: Is the name of this lint ok. it seems to mostly be fine with [convention](https://github.com/rust-lang/rfcs/blob/cc53afbe5dea41e1f7d1c3dce71e013abe025211/text/0344-conventions-galore.md#lints)
[ ]: I'm not super thrilled about the warning text

r? @kmcallister (Shamelessly nominating because you were looking at my other ticket)

### CODE REVIEW

File: src/librustc/lint/builtin.rs [hunk 1]
```diff
@@ -2083,6 +2090,12 @@ impl LintPass for PrivateNoMangleFns {
                     cx.span_lint(PRIVATE_NO_MANGLE_FNS, it.span, msg.as_slice());
                 }
             },
+            ast::ItemConst(..) => {
+                if attr::contains_name(it.attrs.as_slice(), "no_mangle") {
+                    let msg = format!("const items cannot export symbols. {} therefore cannot be #[no_mangle",
```
Comment (richo) [inline src/librustc/lint/builtin.rs]: And I guess this line needs to be broken up to appease tidy

**Comment by richo** [comment]
I rethought this a bit. I don't think there's ever a reason to #[no_mangle] a const, and private #[no_mangle]'d statics should throw a warning as well.

I added the new lint, and tests for all.

It's possible to reuse some code between the Static and Fn check, but it seemed liek it would be way less readable so I let it be duplicated. I think this is now good to go out

**Comment by richo** [comment]
Aaaaand confirmed that the tests still pass. r? @kmcallister

File: src/librustc/lint/builtin.rs [hunk 2]
```diff
@@ -2083,6 +2097,20 @@ impl LintPass for PrivateNoMangleFns {
...
+                if attr::contains_name(it.attrs.as_slice(), "no_mangle") &&
+                       !cx.exported_items.contains(&it.id) {
+                    let msg = format!("static {} is marked #[no_mangle], but not exported",
+                                      it.ident);
+                    cx.span_lint(PRIVATE_NO_MANGLE_STATICS, it.span, msg.as_slice());
+                }
+            },
+            ast::ItemConst(..) => {
+                if attr::contains_name(it.attrs.as_slice(), "no_mangle") {
+                    let msg = "const items should never be #[no_mangle]";
```
Comment (kmcallister) [inline src/librustc/lint/builtin.rs]: Can you add a few words about why?

Comment (richo) [inline src/librustc/lint/builtin.rs]: Sure. How concerned should I be about overly long messages? (The current
ones all seem pretty terse).

I think this should probably have a hint that "maybe you wanted pub static"
too

**Comment by richo** [comment]
I added another commit with a comment and a help message about the thing a user probably wanted

File: src/librustc/lint/builtin.rs [hunk 3]
```diff
@@ -2083,6 +2097,23 @@ impl LintPass for PrivateNoMangleFns {
...
+                    cx.span_lint(PRIVATE_NO_MANGLE_STATICS, it.span, msg.as_slice());
+                }
+            },
+            ast::ItemConst(..) => {
+                if attr::contains_name(it.attrs.as_slice(), "no_mangle") {
+                    // Const items do not refer to a particular location in memory, and therefore
+                    // don't have anything to attach a symbol to
+                    let msg = "const items should never be #[no_mangle]";
+                    cx.span_lint(NO_MANGLE_CONST_ITEMS, it.span, msg);
+                    cx.sess().span_help(it.span, "maybe you wanted this to be a pub static");
```
Comment (kmcallister) [inline src/librustc/lint/builtin.rs]: Unfortunately the help message will print even with `-A no-mangle-const-items`, see #19668

Comment (richo) [inline src/librustc/lint/builtin.rs]: I'm not sure that's a huge issue in this case. I'm not aware of any time
you would actually want this.

Comment (kmcallister) [inline src/librustc/lint/builtin.rs]: Still, the lint should behave consistently.

**Comment by richo** [comment]
Turned the warning + hint into one longer warning.

**Pull request merged**